x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -176,32 +334,39 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
         MappedFieldType idField = context.getFieldType(IdFieldMapper.NAME);
-        if (idField == null) {
+        MappedFieldType indexField = context.getFieldType(IndexFieldMapper.NAME);
+        if (idField == null || indexField == null) {
             return new MatchNoDocsQuery("No mappings");
         }
-        if (this.ids.isEmpty()) {
+        List<Item> items = (documents != null) ? documents : ids.stream().map(id -> new Item(null, id)).collect(Collectors.toList());
+        if (items.isEmpty()) {
             return new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE);
         } else {
-            BooleanQuery.Builder pinnedQueries = new BooleanQuery.Builder();
+            List<Query> pinnedQueries = new ArrayList<>();
 
             // Ensure each pin order using a Boost query with the relevant boost factor
             int minPin = NumericUtils.floatToSortableInt(MAX_ORGANIC_SCORE) + 1;
-            int boostNum = minPin + ids.size();
+            int boostNum = minPin + items.size();
             float lastScore = Float.MAX_VALUE;
-            for (String id : ids) {
+            for (Item item : items) {
                 float pinScore = NumericUtils.sortableIntToFloat(boostNum);
                 assert pinScore < lastScore;
                 lastScore = pinScore;
                 boostNum--;
+                BooleanQuery.Builder documentQueries = new BooleanQuery.Builder();
+                documentQueries.add(idField.termQuery(item.id, context), BooleanClause.Occur.FILTER);
+                if (item.index != null) {
+                    documentQueries.add(indexField.termQuery(item.index, context), BooleanClause.Occur.FILTER);
+                }
                 // Ensure the pin order using a Boost query with the relevant boost factor
-                Query idQuery = new BoostQuery(new ConstantScoreQuery(idField.termQuery(id, context)), pinScore);
-                pinnedQueries.add(idQuery, BooleanClause.Occur.SHOULD);
+                Query documentQuery = new BoostQuery(new ConstantScoreQuery(documentQueries.build()), pinScore);
+                pinnedQueries.add(documentQuery);
             }
 
             // Score for any pinned query clause should be used, regardless of any organic clause score, to preserve pin order.
             // Use dismax to always take the larger (ie pinned) of the organic vs pinned scores
             List<Query> organicAndPinned = new ArrayList<>();
-            organicAndPinned.add(pinnedQueries.build());
+            organicAndPinned.add(new DisjunctionMaxQuery(pinnedQueries, 0));
```

Comment:
Is there a motivation for switching from boolean to dismax query? I think the scoring should be identical between the two.

Reply:
Yeah this got a bit lost in the messy commits: 7d98d38ed853bf05db2b0d4a117c8f586c0b7b2e

I'm not entirely sure, but it seemed like `should` was combining scores in a way that caused issues when multiple `documents` clauses matched the same document.

Reply:
Oh I see, I didn't realize we now allow duplicate IDs in the list. I wonder if we should just require `_index` to always be set, to keep the behavior as simple as possible. I can't think of a use case for mixing both ID-only with ID-plus-index items -- it seems like you'd always want one or the other?

Reply:
I'm not sure of a use case either. I was thinking someone might want a specific document (`_index` and `_id`) as the top result and then documents with the same id in other indices after that. If that doesn't seem helpful, requiring `_index` sounds ok to me too.